File: framework/backend/go/k1s0-domain-event/event.go
// Package domainevent provides domain event publishing, subscribing, and outbox pattern.
package domainevent

// DomainEvent is the base interface for all domain events.
type DomainEvent interface {
	// EventType returns the event type identifier (e.g., "order.created").
	EventType() string

	// AggregateID returns the aggregate root ID (optional).
	AggregateID() string

	// AggregateType returns the aggregate type (optional).
	AggregateType() string
}

// BaseDomainEvent provides a default implementation of DomainEvent.
type BaseDomainEvent struct {
	Type      string
	AggrID    string
	AggrType  string
}

func (e *BaseDomainEvent) EventType() string     { return e.Type }
func (e *BaseDomainEvent) AggregateID() string    { return e.AggrID }
func (e *BaseDomainEvent) AggregateType() string  { return e.AggrType }

File: framework/backend/python/packages/k1s0-domain-event/src/k1s0_domain_event/event.py
"""Domain event abstract base class."""

from __future__ import annotations

from abc import ABC, abstractmethod


class DomainEvent(ABC):
    """Abstract base class for all domain events.

    Subclasses must implement event_type, aggregate_id, and aggregate_type
    to identify the event and the aggregate it belongs to.
    """

    @abstractmethod
    def event_type(self) -> str:
        """Return the unique type identifier for this event."""

    @abstractmethod
    def aggregate_id(self) -> str:
        """Return the ID of the aggregate that produced this event."""

    @abstractmethod
    def aggregate_type(self) -> str:
        """Return the type of the aggregate that produced this event."""
